fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedCompoundReader.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * LuceneOptimizedCompoundReader.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.util.IOUtils;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ *
+ * Borrowed from Lucene to remove checksum from end of file.
```

Comment:
What's the issue with having checksums? Is there a performance penalty beyond the CPU to compute and the extra bytes to store the checksum? (E.g., does it introduce extra FDB transaction conflicts or something like that.)

Reply:
My understanding is the main motivation is to get rid of the unnecessary checksum computation which is quite expensive, based on the assumption that exceptions are thrown at higher level if the content of the file is corrupted.

Reply:
I see. I guess the proof of the pudding is in the eating, but I'm a little skeptical that the CPU work for the checksum is a huge optimization unless it also somehow lets us remove a read (or decrease the amount of pages we need to rewrite or something). But it's maybe worth a shot to see how it goes

Reply:
The problem is that the checksum is stored at the end of the file.  We chunk the file into blocks and this requires us to read an extra block (~ 16K).

Reply:
I see; that makes sense. It would probably be good to add this to the Javadoc, then, as it's a little subtle and easily missed

Reply:
Added one line in the javadoc: ```This implementation skips the reading and validation of checksum, which could have quite big data size sometimes.```